landoapi/api/try.py

```diff
@@ -0,0 +1,77 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import base64
+import logging
+
+from connexion import ProblemException
+from flask import (
+    current_app,
+    g,
+)
+
+from landoapi import auth
+from landoapi.models.landing_job import (
+    LandingJobStatus,
+    add_job_with_revisions,
+)
+from landoapi.models.revisions import Revision
+from landoapi.repos import get_repos_for_env
+
+logger = logging.getLogger(__name__)
+
+
+@auth.require_auth0(scopes=("lando", "profile", "email"), userinfo=True)
+@auth.enforce_scm_level_1
+def post(data: dict):
+    base_commit = data["base_commit"]
+    patches = data["patches"]
+
+    if not base_commit or len(base_commit) != 40:
+        raise ProblemException(
+            400,
+            "Base commit must be a 40-character commit hash.",
+            "Base commit must be a 40-character commit hash.",
+            type="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400",
+        )
+
+    if not patches:
+        raise ProblemException(
+            400,
+            "Patches must contain at least 1 patch.",
+            "Patches must contain at least 1 patch.",
+            type="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400",
+        )
+
+    try_repo = get_repos_for_env(current_app.config.get("ENVIRONMENT")).get("try")
+    if not try_repo:
+        raise ProblemException(
+            500,
+            "Could not find a `try` repo to submit to.",
+            "Could not find a `try` repo to submit to.",
+            type="https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/500",
+        )
+
+    # Add a landing job for this try push.
+    ldap_username = g.auth0_user.email
+    revisions = [
+        # TODO do something more useful with `patch_data`, maybe gather data from
+        # rs-parsepatch??
+        Revision(patch_bytes=base64.b64decode(patch.encode("ascii")), patch_data={})
```

Comment:
This should ideally use `Revision.set_patch` instead, which already encodes `patch` (perhaps you can pass the encoding to that method if it should be a different one) and it uses `patch_data` when calling `build_patch_for_revision`. You can pass boilerplate `patch_data`if needed, but it is better to have this go through the conventional process of patch setting rather than directly setting the patch here.

Another reason to do that apart from the above, and reducing repetition is that in the future, there may be some restrictions as to when and how `patch_bytes` can be set -- for example, if a request to land is issued, there should be no way to modify `patch_bytes`. It would be good to keep this convention and use `set_patch` just in case any other checks or restrictions are needed in the future.

Reply:
My plan was to have the client simply `hg export` the patches so they could be applied directly. `Revision.set_patch` builds an `hg export` formatted patch from a diff and metadata, which might end up being unnecessary and slower to parse from patches on the client. I'm still tinkering with the best way to get patch data into the request.

Reply:
> My plan was to have the client simply `hg export` the patches so they could be applied directly. `Revision.set_patch` builds an `hg export` formatted patch from a diff and metadata, which might end up being unnecessary and slower to parse from patches on the client. I'm still tinkering with the best way to get patch data into the request.

If you want to apply them as-is, then I would still go through `Revision.set_patch` but modify it so that it can take a full patch and circumvent the need for extra patch data. That way we would still be able to apply any future changes to when/how patches can be modified (e.g., a status check or a hash check). In the future we may want to do a pre-update hook instead but that is out of scope for now.

Reply:
One other thing to keep in mind is the patch format expectations that the landing worker has (e.g., `Diff Start Line` header, etc.) which may not be as expected if using the diff as-is.